pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test: TestSQLStatsCompactor failed

pkg/sql/sqlstats/persistedsqlstats/persistedsqlstats_test.TestSQLStatsCompactor [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14474849?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14474849?buildTab=artifacts#/) on master @ [0e6deab01e116667459ba4132a3e6d7c5cd5cbe5](https://github.com/cockroachdb/cockroach/commits/0e6deab01e116667459ba4132a3e6d7c5cd5cbe5):


```
=== RUN   TestSQLStatsCompactor
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/da69a73e19420a8109697ab7be6f5bb3/logTestSQLStatsCompactor109822825
    test_log_scope.go:81: use -show-logs to present logs inline
    compaction_test.go:284: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/da69a73e19420a8109697ab7be6f5bb3/logTestSQLStatsCompactor109822825
--- FAIL: TestSQLStatsCompactor (42.08s)
=== RUN   TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=2
    test_server_shim.go:157: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    compaction_test.go:253: expected 32 number of wide scans issued, but 33 number of wide scan issued
        details: [wide scan in Scan [/Tenant/10/Table/42/10/0,/Tenant/10/Table/42/10/1), [txn: a4b0fb2c], [max_span_request_keys: 0], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/42/10/0,/Tenant/10/Table/42/10/0/2024-03-20T15:59:59.999999001Z), [txn: 0fb78906], [can-forward-ts], [max_span_request_keys: 2], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/42/10/1,/Tenant/10/Table/42/10/2), [txn: d088e54e], [max_span_request_keys: 0], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/42/10/1,/Tenant/10/Table/42/10/1/2024-03-20T15:59:59.999999001Z), [txn: 4d84ec78], [can-forward-ts], [max_span_request_keys: 2], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/42/10/2,/Tenant/10/Table/42/10/3), [txn: 069dd345], [max_span_request_keys: 0], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/42/10/2,/Tenant/10/Table/42/10/2/2024-03-20T15:59:59.999999001Z), [txn: dd3422ef], [can-forward-ts], [max_span_request_keys: 2], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/42/10/3,/Tenant/10/Table/42/10/4), [txn: ffbbb0f2], [max_span_request_keys: 0], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/42/10/3,/Tenant/10/Table/42/10/3/2024-03-20T15:59:59.999999001Z), [txn: a55310c0], [can-forward-ts], [max_span_request_keys: 2], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/42/10/4,/Tenant/10/Table/42/10/5), [txn: d60f9a07], [max_span_request_keys: 0], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/42/10/4,/Tenant/10/Table/42/10/4/2024-03-20T15:59:59.999999001Z), [txn: 82463806], [can-forward-ts], [max_span_request_keys: 2], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/42/10/5,/Tenant/10/Table/42/10/6), [txn: 82d36fee], [max_span_request_keys: 0], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/42/10/5,/Tenant/10/Table/42/10/5/2024-03-20T15:59:59.999999001Z), [txn: 30b9ae3c], [can-forward-ts], [max_span_request_keys: 2], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/42/10/6,/Tenant/10/Table/42/10/7), [txn: 362aaa5d], [max_span_request_keys: 0], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/42/10/6,/Tenant/10/Table/42/10/6/2024-03-20T15:59:59.999999001Z), [txn: 362bd2cc], [can-forward-ts], [max_span_request_keys: 2], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/42/10/7,/Tenant/10/Table/42/10/8), [txn: ce143013], [max_span_request_keys: 0], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/42/10/7,/Tenant/10/Table/42/10/7/2024-03-20T15:59:59.999999001Z), [txn: 3dbd5856], [can-forward-ts], [max_span_request_keys: 2], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/43/9/0,/Tenant/10/Table/43/9/1), [txn: 484530e5], [max_span_request_keys: 0], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/43/9/0,/Tenant/10/Table/43/9/0/2024-03-20T15:59:59.999999001Z), [txn: 2452058c], [can-forward-ts], [max_span_request_keys: 2], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/43/9/1,/Tenant/10/Table/43/9/2), [txn: bdb06807], [max_span_request_keys: 0], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/43/9/1,/Tenant/10/Table/43/9/1/2024-03-20T15:59:59.999999001Z), [txn: 17deaeb9], [can-forward-ts], [max_span_request_keys: 2], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/43/9/1,/Tenant/10/Table/43/9/1/2024-03-20T15:59:59.999999001Z), [txn: 17deaeb9], [can-forward-ts], [max_span_request_keys: 2], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/43/9/2,/Tenant/10/Table/43/9/3), [txn: 5c05276f], [max_span_request_keys: 0], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/43/9/2,/Tenant/10/Table/43/9/2/2024-03-20T15:59:59.999999001Z), [txn: 69d2b6d3], [can-forward-ts], [max_span_request_keys: 2], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/43/9/3,/Tenant/10/Table/43/9/4), [txn: bb06791c], [max_span_request_keys: 0], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/43/9/3,/Tenant/10/Table/43/9/3/2024-03-20T15:59:59.999999001Z), [txn: 04189006], [can-forward-ts], [max_span_request_keys: 2], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/43/9/4,/Tenant/10/Table/43/9/5), [txn: 05e1b923], [max_span_request_keys: 0], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/43/9/4,/Tenant/10/Table/43/9/4/2024-03-20T15:59:59.999999001Z), [txn: 5e53d256], [can-forward-ts], [max_span_request_keys: 2], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/43/9/5,/Tenant/10/Table/43/9/6), [txn: bcd2e090], [max_span_request_keys: 0], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/43/9/5,/Tenant/10/Table/43/9/5/2024-03-20T15:59:59.999999001Z), [txn: d97df920], [can-forward-ts], [max_span_request_keys: 2], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/43/9/6,/Tenant/10/Table/43/9/7), [txn: 147904c6], [max_span_request_keys: 0], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/43/9/6,/Tenant/10/Table/43/9/6/2024-03-20T15:59:59.999999001Z), [txn: f2c5af11], [can-forward-ts], [max_span_request_keys: 2], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/43/9/7,/Tenant/10/Table/43/9/8), [txn: 8650fe9d], [max_span_request_keys: 0], [target_bytes: 10485760] wide scan in Scan [/Tenant/10/Table/43/9/7,/Tenant/10/Table/43/9/7/2024-03-20T15:59:59.999999001Z), [txn: 0106446b], [can-forward-ts], [max_span_request_keys: 2], [target_bytes: 10485760]]
    --- FAIL: TestSQLStatsCompactor/stmtCount=200/maxPersistedRowLimit=40/rowsDeletePerTxn=2 (6.82s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/obs-inf-prs
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSQLStatsCompactor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/server/application_api/sql_stats_test.go
+++ b/pkg/server/application_api/sql_stats_test.go
@@ -1345,7 +1345,7 @@ func TestCombinedStatementUsesCorrectSourceTable(t *testing.T) {
 		stmt.AggregatedTs = startTs
 		stmt.Key.App = server.CrdbInternalStmtStatsPersisted
 		stmt.Key.TransactionFingerprintID = 1
-		require.NoError(t, sqlstatstestutil.InsertMockedIntoSystemStmtStats(ctx, ie, &stmt, 1 /* nodeId */, nil))
+		require.NoError(t, sqlstatstestutil.InsertMockedIntoSystemStmtStats(ctx, ie, []appstatspb.CollectedStatementStatistics{stmt}, 1))
 
 		stmt.Key.App = server.CrdbInternalStmtStatsCached
 		require.NoError(t, sqlstatstestutil.InsertMockedIntoSystemStmtActivity(ctx, ie, &stmt, nil))
@@ -1355,7 +1355,7 @@ func TestCombinedStatementUsesCorrectSourceTable(t *testing.T) {
 		txn.TransactionFingerprintID = 1
 		txn.AggregatedTs = startTs
 		txn.App = server.CrdbInternalTxnStatsPersisted
-		require.NoError(t, sqlstatstestutil.InsertMockedIntoSystemTxnStats(ctx, ie, &txn, 1, nil))
+		require.NoError(t, sqlstatstestutil.InsertMockedIntoSystemTxnStats(ctx, ie, []appstatspb.CollectedTransactionStatistics{txn}, 1))
 		txn.App = server.CrdbInternalTxnStatsCached
 		require.NoError(t, sqlstatstestutil.InsertMockedIntoSystemTxnActivity(ctx, ie, &txn, nil))
 

--- a/pkg/sql/sqlstats/persistedsqlstats/BUILD.bazel
+++ b/pkg/sql/sqlstats/persistedsqlstats/BUILD.bazel
@@ -86,7 +86,6 @@ go_test(
         "//pkg/scheduledjobs",
         "//pkg/security/securityassets",
         "//pkg/security/securitytest",
-        "//pkg/security/username",
         "//pkg/server",
         "//pkg/server/serverpb",
         "//pkg/sql",

--- a/pkg/sql/sqlstats/persistedsqlstats/compaction_test.go
+++ b/pkg/sql/sqlstats/persistedsqlstats/compaction_test.go
@@ -8,6 +8,7 @@ package persistedsqlstats_test
 import (
 	"context"
 	"fmt"
+	"math"
 	"regexp"
 	"sync/atomic"
 	"testing"
@@ -20,17 +21,15 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
-	"github.com/cockroachdb/cockroach/pkg/security/username"
-	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/cockroach/pkg/sql/appstatspb"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
 	"github.com/cockroachdb/cockroach/pkg/sql/isql"
-	"github.com/cockroachdb/cockroach/pkg/sql/sem/catconstants"
 	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlstats"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats/sqlstatstestutil"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/encoding"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -38,6 +37,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/metric"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/cockroach/pkg/util/uuid"
 	"github.com/stretchr/testify/require"
 )
 
@@ -69,8 +69,6 @@ func TestSQLStatsCompactorNilTestingKnobCheck(t *testing.T) {
 func TestSQLStatsCompactor(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
-
-	skip.WithIssue(t, 120761)
 
 	ctx := context.Background()
 
@@ -126,12 +124,13 @@ func TestSQLStatsCompactor(t *testing.T) {
 			tc.maxPersistedRowLimit,
 			tc.rowsToDeletePerTxn,
 		), func(t *testing.T) {
-			kvInterceptor := kvScanInterceptor{}
+			kvInterceptor := newKVScanIntercceptor()
 			cleanupInterceptor := cleanupInterceptor{}
 
+			stubSqlStatsTime := timeutil.Now().Add(-2 * time.Hour).Truncate(time.Hour)
 			knobs := sqlstats.CreateTestingKnobs()
 			knobs.StubTimeNow = func() time.Time {
-				return timeutil.Now().Add(-2 * time.Hour)
+				return stubSqlStatsTime
 			}
 			srv, conn, _ := serverutils.StartServer(
 				t, base.TestServerArgs{
@@ -151,26 +150,12 @@ func TestSQLStatsCompactor(t *testing.T) {
 			internalExecutor := server.InternalExecutor().(isql.Executor)
 
 			func() {
-				// Determine the actual ID of the stat tables.
-				sID, err := server.QueryTableID(ctx, username.RootUserName(),
-					"system", string(catconstants.StatementStatisticsTableName))
-				require.NoError(t, err)
-				tID, err := server.QueryTableID(ctx, username.RootUserName(),
-					"system", string(catconstants.TransactionStatisticsTableName))
-				require.NoError(t, err)
-
-				// Configure the KV interceptor. We also need the codec from
-				// the live server, which determines the tenant ID prefix to
-				// strip.
-				kvInterceptor.mu.Lock()
-				defer kvInterceptor.mu.Unlock()
-
-				kvInterceptor.mu.codec = server.Codec()
-				kvInterceptor.mu.stmtStatsTableID = uint32(sID)
-				kvInterceptor.mu.txnStatsTableID = uint32(tID)
+				// Set the codec from the live server, which determines the tenant
+				// ID prefix to strip.
+				kvInterceptor.codec = server.Codec()
 			}()
 
-			// Disable automatic flush since the test will handle the flush manually.
+			// Disable automatic flush since the test will insert mocked data manually.
 			sqlConn.Exec(t, "SET CLUSTER SETTING sql.stats.flush.interval = '24h'")
 			// Disable activity update flush which also does a scan on the stats table
 			sqlConn.Exec(t, "SET CLUSTER SETTING sql.stats.activity.flush.enabled = false")
@@ -194,10 +179,6 @@ func TestSQLStatsCompactor(t *testing.T) {
 				"TRUNCATE system.transaction_statistics",
 			)
 			require.NoError(t, err)
-			serverSQLStats :=
-				server.
-					SQLServer().(*sql.Server).
-					GetSQLStatsProvider().(*persistedsqlstats.PersistedSQLStats)
 			sqlConn.Exec(t,
 				"SET CLUSTER SETTING sql.stats.persisted_rows.max = $1",
 				tc.maxPersistedRowLimit)
@@ -210,8 +191,17 @@ func TestSQLStatsCompactor(t *testing.T) {
 				sqlConn.Exec(t, "RESET CLUSTER SETTING sql.stats.cleanup.rows_to_delete_per_txn")
 			}
 
-			generateFingerprints(t, sqlConn, tc.stmtCount)
-			serverSQLStats.MaybeFlush(ctx, server.AppStopper())
+			stmts := make([]appstatspb.CollectedStatementStatistics, tc.stmtCount)
+			txns := make([]appstatspb.CollectedTransactionStatistics, tc.stmtCount)
+			for i := 0; i < tc.stmtCount; i++ {
+				stmts[i].ID = appstatspb.StmtFingerprintID(i)
+				stmts[i].Key.TransactionFingerprintID = appstatspb.TransactionFingerprintID(i)
+				stmts[i].AggregatedTs = stubSqlStatsTime
+				txns[i].TransactionFingerprintID = appstatspb.TransactionFingerprintID(i)
+				txns[i].AggregatedTs = stubSqlStatsTime
+			}
+			require.NoError(t, sqlstatstestutil.InsertMockedIntoSystemStmtStats(ctx, internalExecutor, stmts, 1))
+			require.NoError(t, sqlstatstestutil.InsertMockedIntoSystemTxnStats(ctx, internalExecutor, txns, 1))
 
 			sqlStatsKnobs := sqlstats.CreateTestingKnobs()
 			sqlStatsKnobs.OnCleanupStartForShard = cleanupInterceptor.intercept
@@ -226,9 +216,10 @@ func TestSQLStatsCompactor(t *testing.T) {
 			expectedDeletedStmtFingerprints, expectedDeletedTxnFingerprints :=
 				getTopSortedFingerprints(t, sqlConn, tc.maxPersistedRowLimit)
 
+			expectedCount := int(math.Min(float64(tc.stmtCount), float64(tc.maxPersistedRowLimit)))
 			// Sanity check.
-			require.Equal(t, tc.maxPersistedRowLimit, len(expectedDeletedStmtFingerprints))
-			require.Equal(t, tc.maxPersistedRowLimit, len(expectedDeletedTxnFingerprints))
+			require.Equal(t, expectedCount, len(expectedDeletedStmtFingerprints))
+			require.Equal(t, expectedCount, len(expectedDeletedTxnFingerprints))
 
 			// The two interceptors (kvInterceptor and cleanupInterceptor) are
 			// injected into kvserver and StatsCompactor respectively.
@@ -256,7 +247,8 @@ func TestSQLStatsCompactor(t *testing.T) {
 			actualStmtFingerprints, actualTxnFingerprints :=
 				getTopSortedFingerprints(t, sqlConn, 0 /* limit */)
 
-			require.GreaterOrEqual(t, tc.maxPersistedRowLimit, len(actualStmtFingerprints))
+			require.GreaterOrEqual(t, expectedCount, len(actualStmtFingerprints))
+			require.GreaterOrEqual(t, expectedCount, len(actualTxnFingerprints))
 			require.GreaterOrEqual(t, tc.maxPersistedRowLimit, len(actualTxnFingerprints))
 
 			for fingerprintID := range actualStmtFingerprints {
@@ -275,8 +267,8 @@ func TestSQLStatsCompactor(t *testing.T) {
 			err = statsCompactor.DeleteOldestEntries(ctx)
 			require.NoError(t, err)
 			stmtStatsCnt, txnStatsCnt := getPersistedStatsEntry(t, sqlConn)
-			require.GreaterOrEqual(t, tc.maxPersistedRowLimit, stmtStatsCnt)
-			require.GreaterOrEqual(t, tc.maxPersistedRowLimit, txnStatsCnt)
+			require.GreaterOrEqual(t, expectedCount, stmtStatsCnt)
+			require.GreaterOrEqual(t, expectedCount, txnStatsCnt)
 		})
 	}
 }
@@ -380,14 +372,10 @@ ORDER BY aggregated_ts`
 	return stmtFingerprints, txnFingerprints
 }
 
-func generateFingerprints(t *testing.T, sqlConn *sqlutils.SQLRunner, distinctFingerprints int) {
-	stmt := "SELECT 1"
-	for i := 0; i < distinctFingerprints; i++ {
-		sqlConn.Exec(t, stmt)
-		// Mutate the stmt to create different fingerprint.
-		stmt = fmt.Sprintf("%s, 1", stmt)
-	}
-}
+const (
+	StatementStatisticsTableID   = 42
+	TransactionStatisticsTableID = 43
+)
 
 // Tables 42 and 43 and are the IDs of the statement and transactions
 // tables respectively. See `StatementStatisticsTableID` and
@@ -395,26 +383,34 @@ func generateFingerprints(t *testing.T, sqlConn *sqlutils.SQLRunner, distinctFin
 var kvReqWideScanStartKeyPattern = regexp.MustCompile(`(/Tenant/\d+)?/Table/((42)|(43))/[0-9]{1,2}/[0-9]$`)
 
 type kvScanInterceptor struct {
-	totalWideScan   int64
+	totalWideScan   atomic.Int64
 	enabled         int32
 	wideScanDetails []string
 
+	// The codec must be set before the interceptor is enabled.
+	codec keys.SQLCodec
+
 	mu struct {
 		syncutil.Mutex
-
-		stmtStatsTableID uint32
-		txnStatsTableID  uint32
-		codec            keys.SQLCodec
+		// Track the txnIDs of the scans since
+		// the txn could be retried.
+		txnIDs map[uuid.UUID]struct{}
 	}
 }
 
+func newKVScanIntercceptor() *kvScanInterceptor {
+	k := kvScanInterceptor{}
+	k.mu.txnIDs = make(map[uuid.UUID]struct{})
+	return &k
+}
+
 func (k *kvScanInterceptor) reset() {
-	atomic.StoreInt64(&k.totalWideScan, 0)
+	k.totalWideScan.Store(0)
 	k.wideScanDetails = []string{}
 }
 
 func (k *kvScanInterceptor) getTotalWideScans() int64 {
-	return atomic.LoadInt64(&k.totalWideScan)
+	return k.totalWideScan.Load()
 }
 
 func (k *kvScanInterceptor) enable() {
@@ -429,42 +425,63 @@ func (k *kvScanInterceptor) intercept(ctx context.Context, ba *kvpb.BatchRequest
 	if atomic.LoadInt32(&k.enabled) == 0 {
 		return nil
 	}
-	if req, ok := ba.GetArg(kvpb.Scan); ok {
-		k.mu.Lock()
-		defer k.mu.Unlock()
-		_, tableID, err := k.mu.codec.DecodeTablePrefix(req.(*kvpb.ScanRequest).Key)
-		if err != nil {
-			log.Warningf(ctx, "unable to decode prefix: %v", err)
-		}
-		if tableID == k.mu.stmtStatsTableID || tableID == k.mu.txnStatsTableID {
-			prettyKey := roachpb.PrettyPrintKey([]encoding.Direction{}, req.(*kvpb.ScanRequest).Key)
 
-			keyMatchedWideScan := kvReqWideScanStartKeyPattern.MatchString(prettyKey)
+	req, ok := ba.GetArg(kvpb.Scan)
+	if !ok {
+		return nil
+	}
 
-			if keyMatchedWideScan {
-				k.wideScanDetails = append(k.wideScanDetails, fmt.Sprintf("wide scan in %v", ba))
-				atomic.AddInt64(&k.totalWideScan, 1)
-			}
+	_, tableID, err := k.codec.DecodeTablePrefix(req.(*kvpb.ScanRequest).Key)
+	if err != nil {
+		log.Warningf(ctx, "unable to decode prefix: %v", err)
+	}
+
+	// Ensure the request is for one of the sql stats tables.
+	if tableID != StatementStatisticsTableID && tableID != TransactionStatisticsTableID {
+		return nil
+	}
+
+	k.mu.Lock()
+	defer k.mu.Unlock()
+	// Avoid double counting the same txn.
+	if ba.Txn != nil {
+		txnID := ba.Txn.ID
+		if _, ok := k.mu.txnIDs[txnID]; ok {
+			return nil
 		}
+		k.mu.txnIDs[txnID] = struct{}{}
+	}
+
+	prettyKey := roachpb.PrettyPrintKey([]encoding.Direction{}, req.(*kvpb.ScanRequest).Key)
+	keyMatchedWideScan := kvReqWideScanStartKeyPattern.MatchString(prettyKey)
+
+	if keyMatchedWideScan {
+		k.wideScanDetails = append(k.wideScanDetails, fmt.Sprintf("wide scan in %v", ba))
+		k.totalWideScan.Add(1)
 	}
 
 	return nil
 }
 
 type cleanupInterceptor struct {
-	expectedNumberOfWideScans int64
+	expectedNumberOfWideScans atomic.Int64
 }
 
 func (c *cleanupInterceptor) reset() {
-	atomic.StoreInt64(&c.expectedNumberOfWideScans, 0)
+	c.expectedNumberOfWideScans.Store(0)
 }
 
-func (c *cleanupInterceptor) intercept(shardIdx int, existingCountInShard, shardLimit int64) {
+func (c *cleanupInterceptor) intercept(_ int, existingCountInShard, shardLimit int64) {
 	if existingCountInShard > shardLimit {
-		atomic.AddInt64(&c.expectedNumberOfWideScans, 1)
+		c.expectedNumberOfWideScans.Add(1)
 	}
 }
 
 func (c *cleanupInterceptor) getExpectedNumberOfWideScans() int64 {
-	return systemschema.SQLStatsHashShardBucketCount*2 + atomic.LoadInt64(&c.expectedNumberOfWideScans)
+	// The first part of the equation is the number of wide scans that
+	// result from getting the shard count for each shard in each table.
+	// The second part is the number of wide scans that result from the
+	// delete statement, which is only executed if the number of rows
+	// in the shard exceeds the shard limit.
+	return systemschema.SQLStatsHashShardBucketCount*2 + c.expectedNumberOfWideScans.Load()
 }

--- a/pkg/sql/sqlstats/persistedsqlstats/sqlstatstestutil/BUILD.bazel
+++ b/pkg/sql/sqlstats/persistedsqlstats/sqlstatstestutil/BUILD.bazel
@@ -9,8 +9,10 @@ go_library(
         "//pkg/base",
         "//pkg/sql",
         "//pkg/sql/appstatspb",
+        "//pkg/sql/isql",
         "//pkg/sql/sem/tree",
         "//pkg/sql/sessiondata",
         "//pkg/sql/sqlstats/persistedsqlstats/sqlstatsutil",
+        "//pkg/util/safesql",
     ],
 )

--- a/pkg/sql/sqlstats/persistedsqlstats/sqlstatstestutil/testutils.go
+++ b/pkg/sql/sqlstats/persistedsqlstats/sqlstatstestutil/testutils.go
@@ -14,9 +14,11 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/sql"
 	"github.com/cockroachdb/cockroach/pkg/sql/appstatspb"
+	"github.com/cockroachdb/cockroach/pkg/sql/isql"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats/sqlstatsutil"
+	"github.com/cockroachdb/cockroach/pkg/util/safesql"
 )
 
 // GetRandomizedCollectedStatementStatisticsForTest returns a
@@ -43,52 +45,67 @@ func GetRandomizedCollectedTransactionStatisticsForTest(
 
 func InsertMockedIntoSystemStmtStats(
 	ctx context.Context,
-	ie *sql.InternalExecutor,
-	stmtStats *appstatspb.CollectedStatementStatistics,
+	ie isql.Executor,
+	stmtStatsList []appstatspb.CollectedStatementStatistics,
 	nodeID base.SQLInstanceID,
-	aggInterval *time.Duration,
 ) error {
-	if stmtStats == nil {
+	if len(stmtStatsList) == 0 {
 		return nil
 	}
 
 	aggIntervalVal := time.Hour
-	if aggInterval != nil {
-		aggIntervalVal = *aggInterval
+
+	// Initialize the query builder
+	query := safesql.NewQuery()
+	query.Append("UPSERT INTO system.statement_statistics ")
+	query.Append("(aggregated_ts, fingerprint_id, transaction_fingerprint_id, plan_hash, app_name, node_id, agg_interval, metadata, statistics, plan) VALUES ")
+
+	for i, stmtStats := range stmtStatsList {
+		if i > 0 {
+			query.Append(", ")
+		}
+
+		stmtFingerprint := sqlstatsutil.EncodeUint64ToBytes(uint64(stmtStats.ID))
+		txnFingerprint := sqlstatsutil.EncodeUint64ToBytes(uint64(stmtStats.Key.TransactionFingerprintID))
+		planHash := sqlstatsutil.EncodeUint64ToBytes(stmtStats.Key.PlanHash)
+
+		metadataJSON, err := sqlstatsutil.BuildStmtMetadataJSON(&stmtStats)
+		if err != nil {
+			return err
+		}
+
+		statisticsJSON, err := sqlstatsutil.BuildStmtStatisticsJSON(&stmtStats.Stats)
+		if err != nil {
+			return err
+		}
+		statistics := tree.NewDJSON(statisticsJSON)
+
+		plan := tree.NewDJSON(sqlstatsutil.ExplainTreePlanNodeToJSON(&stmtStats.Stats.SensitiveInfo.MostRecentPlanDescription))
+
+		metadata := tree.NewDJSON(metadataJSON)
+
+		query.Append("($, $, $, $, $, $, $, $, $, $)",
+			stmtStats.AggregatedTs, // aggregated_ts
+			stmtFingerprint,        // fingerprint_id
+			txnFingerprint,         // transaction_fingerprint_id
+			planHash,               // plan_hash
+			stmtStats.Key.App,      // app_name
+			nodeID,                 // node_id
+			aggIntervalVal,         // agg_interval
+			metadata,               // metadata
+			statistics,             // statistics
+			plan,                   // plan
+		)
 	}
 
-	stmtFingerprint := sqlstatsutil.EncodeUint64ToBytes(uint64(stmtStats.ID))
-	txnFingerprint := sqlstatsutil.EncodeUint64ToBytes(uint64(stmtStats.Key.TransactionFingerprintID))
-	planHash := sqlstatsutil.EncodeUint64ToBytes(stmtStats.Key.PlanHash)
-
-	metadataJSON, err := sqlstatsutil.BuildStmtMetadataJSON(stmtStats)
-	if err != nil {
-		return err
-	}
-
-	statisticsJSON, err := sqlstatsutil.BuildStmtStatisticsJSON(&stmtStats.Stats)
-	if err != nil {
-		return err
-	}
-	statistics := tree.NewDJSON(statisticsJSON)
-
-	plan := tree.NewDJSON(sqlstatsutil.ExplainTreePlanNodeToJSON(&stmtStats.Stats.SensitiveInfo.MostRecentPlanDescription))
-
-	metadata := tree.NewDJSON(metadataJSON)
-
-	_, err = ie.ExecEx(ctx, "insert-mock-stmt-stats", nil, sessiondata.NodeUserSessionDataOverride,
-		`UPSERT INTO system.statement_statistics
-VALUES ($1 ,$2, $3, $4, $5, $6, $7, $8, $9, $10)`,
-		stmtStats.AggregatedTs, // aggregated_ts
-		stmtFingerprint,        // fingerprint_id
-		txnFingerprint,         // transaction_fingerprint_id
-		planHash,               // plan_hash
-		stmtStats.Key.App,      // app_name
-		nodeID,                 // node_id
-		aggIntervalVal,         // agg_interval
-		metadata,               // metadata
-		statistics,             // statistics
-		plan,                   // plan
+	// Execute the query
+	_, err := ie.ExecEx(
+		ctx,
+		"insert-mock-stmt-stats-batch",
+		nil,
+		sessiondata.NodeUserSessionDataOverride,
+		query.String(),
+		query.QueryArguments()...,
 	)
 
 	return err
@@ -96,47 +113,59 @@ VALUES ($1 ,$2, $3, $4, $5, $6, $7, $8, $9, $10)`,
 
 func InsertMockedIntoSystemTxnStats(
 	ctx context.Context,
-	ie *sql.InternalExecutor,
-	stats *appstatspb.CollectedTransactionStatistics,
+	ie isql.Executor,
+	statsList []appstatspb.CollectedTransactionStatistics,
 	nodeID base.SQLInstanceID,
-	aggInterval *time.Duration,
 ) error {
-	if stats == nil {
+	if len(statsList) == 0 {
 		return nil
 	}
-
 	aggIntervalVal := time.Hour
-	if aggInterval != nil {
-		aggIntervalVal = *aggInterval
+
+	// Initialize the query builder
+	query := safesql.NewQuery()
+	query.Append("UPSERT INTO system.transaction_statistics ")
+	query.Append("(aggregated_ts, fingerprint_id, app_name, node_id, agg_interval, metadata, statistics) VALUES ")
+
+	for i, stats := range statsList {
+		if i > 0 {
+			query.Append(", ")
+		}
+
+		txnFingerprint := sqlstatsutil.EncodeUint64ToBytes(uint64(stats.TransactionFingerprintID))
+
+		statisticsJSON, err := sqlstatsutil.BuildTxnStatisticsJSON(&stats)
+		if err != nil {
+			return err
+		}
+		statistics := tree.NewDJSON(statisticsJSON)
+
+		metadataJSON, err := sqlstatsutil.BuildTxnMetadataJSON(&stats)
+		if err != nil {
+			return err
+		}
+		metadata := tree.NewDJSON(metadataJSON)
+
+		query.Append("($, $, $, $, $, $, $)",
+			stats.AggregatedTs, // aggregated_ts
+			txnFingerprint,     // fingerprint_id
+			stats.App,          // app_name
+			nodeID,             // node_id
+			aggIntervalVal,     // agg_interval
+			metadata,           // metadata
+			statistics,         // statistics
+		)
 	}
 
-	txnFingerprint := sqlstatsutil.EncodeUint64ToBytes(uint64(stats.TransactionFingerprintID))
-
-	statisticsJSON, err := sqlstatsutil.BuildTxnStatisticsJSON(stats)
-	if err != nil {
-		return err
-	}
-	statistics := tree.NewDJSON(statisticsJSON)
-
-	metadataJSON, err := sqlstatsutil.BuildTxnMetadataJSON(stats)
-	if err != nil {
-		return err
-	}
-	metadata := tree.NewDJSON(metadataJSON)
-	aggregatedTs := stats.AggregatedTs
-
-	_, err = ie.ExecEx(ctx, "insert-mock-txn-stats", nil, sessiondata.NodeUserSessionDataOverride,
-		` UPSERT INTO system.transaction_statistics
-VALUES ($1 ,$2, $3, $4, $5, $6, $7)`,
-		aggregatedTs,   // aggregated_ts
-		txnFingerprint, // fingerprint_id
-		stats.App,      // app_name
-		nodeID,         // node_id
-		aggIntervalVal, // agg_interval
-		metadata,       // metadata
-		statistics,     // statistics
+	// Execute the query
+	_, err := ie.ExecEx(
+		ctx,
+		"insert-mock-txn-stats-batch",
+		nil,
+		sessiondata.NodeUserSessionDataOverride,
+		query.String(),
+		query.QueryArguments()...,
 	)
-
 	return err
 }
 

